File: android/app/src/main/java/com/simplespider/dy/data/PlaySrc.kt
package com.simplespider.dy.data

import com.google.gson.annotations.SerializedName

data class PlaySrcSlideDto(
    val kind: String,
    val index: Int,
    val url: String,
)

data class PlaySrcDto(
    @SerializedName("content_type") val contentType: String,
    val video: String? = null,
    val music: String? = null,
    val slides: List<PlaySrcSlideDto>? = null,
)

fun PlaySrcDto?.isPlayable(): Boolean {
    val src = this ?: return false
    if (src.contentType == "video") return !src.video.isNullOrBlank()
    val slides = src.slides.orEmpty()
    return !src.music.isNullOrBlank() || slides.isNotEmpty()
}

fun PlaySrcDto?.isVideoContent(): Boolean =
    this?.contentType == "video" && !this.video.isNullOrBlank()

fun PlaySrcDto?.isSlidesContent(): Boolean {
    val src = this ?: return false
    return src.contentType == "photo_slides" || src.contentType == "video_slides"
}

fun PlaySrcDto?.primaryVideoUrl(): String? =
    if (isVideoContent()) this?.video?.trim()?.takeIf { it.isNotEmpty() } else null

fun PlaySrcDto?.prefetchUrl(): String? {
    val src = this ?: return null
    primaryVideoUrl()?.let { return it }
    src.music?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    return src.slides.orEmpty().firstOrNull()?.url?.trim()?.takeIf { it.isNotEmpty() }
}

fun PlaySrcDto?.avatarSourceUrl(): String? {
    val src = this ?: return null
    primaryVideoUrl()?.let { return it }
    src.music?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
    return src.slides.orEmpty().firstOrNull()?.url?.trim()?.takeIf { it.isNotEmpty() }
}

File: frontend/src/utils/playSrc.js
export function normalizePlaySrc(playSrc) {
  if (!playSrc) return null
  if (typeof playSrc === 'string') {
    return { content_type: 'video', video: playSrc }
  }
  return playSrc
}

export function isPlayable(playSrc) {
  const ps = normalizePlaySrc(playSrc)
  if (!ps) return false
  if (ps.content_type === 'video') return !!ps.video
  const slides = ps.slides || []
  return !!(ps.music || slides.length)
}

export function isSlidesPlaySrc(playSrc) {
  const ps = normalizePlaySrc(playSrc)
  if (!ps) return false
  return ps.content_type === 'photo_slides' || ps.content_type === 'video_slides'
}

export function primaryOpenUrl(playSrc) {
  const ps = normalizePlaySrc(playSrc)
  if (!ps) return null
  if (ps.content_type === 'video') return ps.video || null
  if (ps.music) return ps.music
  const slides = ps.slides || []
  return slides[0]?.url || null
}

export function formatPlaySrcDisplay(playSrc) {
  const ps = normalizePlaySrc(playSrc)
  if (!ps) return '-'
  if (ps.content_type === 'video') return ps.video || '-'
  const lines = []
  if (ps.music) lines.push(`music: ${ps.music}`)
  for (const slide of ps.slides || []) {
    lines.push(`${slide.kind}[${slide.index}]: ${slide.url}`)
  }
  return lines.length ? lines.join('\n') : '-'
}
